thunder/torch/__init__.py

```diff
@@ -5331,6 +5336,223 @@ def wait(slf) -> None:
+
+    if isinstance(arg, (torch.Tensor, FakeTensor, TensorProxy)):
+        return dtypes.is_inexact_dtype(to_dtype(arg.dtype))
+    return False
+
+
+def _make_differentiable_wrapper(func: Callable, *args, **kwargs):
+    flat_args, spec = tree_flatten((args, kwargs))
+    differentiable_args_idx = tuple(i for i, a in enumerate(flat_args) if _is_differentiable(a))
+    differentiable_args = tuple(arg for i, arg in enumerate(flat_args) if i in differentiable_args_idx)
+
+    def wrapper(*diff_args):
+        diff_args_iter = iter(diff_args)
+        full_args = [next(diff_args_iter) if i in differentiable_args_idx else arg for i, arg in enumerate(flat_args)]
+        full_args, full_kwargs = tree_unflatten(full_args, spec)
+        return func(*full_args, **full_kwargs)
+
+    return wrapper, differentiable_args
+
+
+def register_default_torch_ops():
+    from thunder.torch import default_torch_ops
+
+    for m, fns in default_torch_ops.torch_auto_registered_ops.items():
+        for fn in fns:
+            # Ensure no inplace op in the list
+            utils.check(
+                not fn.__name__.endswith("_"),
+                lambda: f"Automatic registration does not support in-place op of {m.__name__}.{fn.__name__}, please manually register it",
+            )
+            register_default_torch_op(fn, meta_adaptor(fn), m)
+
+
+def register_default_torch_op(torchfn: Callable, fn_meta: Callable, m):
+    _fn = langctx(Languages.TORCH)(fn_meta)
+    sym = Symbol(
+        name=torchfn.__name__,
+        meta=_fn,
+        id=f"{m.__name__}.{torchfn.__name__}",
+    )
+    _torch_to_thunder_function_map[torchfn] = sym
+    from thunder.executors.torchex import _always_executable, ex
+
+    op = ex.register_operator(torchfn.__name__, module=m, meta=fn_meta)
+    ex.register_implementation(sym, op, checker=_always_executable)
+
+    from thunder.core.transforms import augmented_forward_impls, backward_impls
+
+    augmented_forward_impls[sym.id] = augmented_forward_adaptor(op)
+
+    def _vjp_impl_wrapper(residules, *gs):
+        return _vjp_impl(torchfn, residules, *gs)
+
+    bwd_op = ex.register_operator(torchfn.__name__ + "_vjp", meta=backward_adaptor(torchfn), fn=_vjp_impl_wrapper)
+    ex.register_implementation(bwd_op.id, bwd_op, checker=_always_executable)
+    backward_impls[sym.id] = bwd_op
+
+
+def _get_fake_arg(inp: Any, fake_mode):
+    if inp is None:
+        return inp
+    if isinstance(inp, NumberProxy):
+        if inp.value == None:
+            raise NotImplementedError("Unsupported for NumberProxy.value=None")
+        else:
+            return inp.value
+    elif isinstance(inp, TensorProxy):
+        return fake_mode.from_tensor(
+            torch.empty(
+                inp.shape,
+                dtype=_thunder_to_torch_dtype_map[inp.dtype],
+                requires_grad=inp.requires_grad,
+                device=devices.to_torch_device(inp.device),
+            )
+        )
```

Comment:
This would unnecessarily reserve device memory that would be freed almost immediately if the fake mode wasn't active. `_get_fake_arg` is called inside the active fake mode, is "from_tensor" necessary? Can we simply return the result of `torch.empty`?

Reply:
Yes, the original intent of passing in fake_mode is to make `_get_fake_arg` a stand along func that can be used outside of the context manager, but since we only use it in side of the active fake mode, I can change it to return the result of torch.empty.